activerecord/lib/active_record/relation.rb

```diff
@@ -857,6 +887,37 @@ def exec_queries(&block)
         end
       end
 
+      def exec_main_query(async: false)
+        skip_query_cache_if_necessary do
+          if where_clause.contradiction?
+            [].freeze
+          elsif eager_loading?
+            apply_join_dependency do |relation, join_dependency|
+              if relation.null_relation?
+                [].freeze
+              else
+                relation = join_dependency.apply_column_aliases(relation)
+                @_join_dependency = join_dependency
```

Comment:
I'm not very happy with having to store this in an instance variable for later use, but `apply_join_dependency` can sometimes trigger a query on its own, so if we were to call it twice, we'd generate an extra query in some cases.

Reply:
There is exactly one test case: https://buildkite.com/rails/rails/builds/74758#40f90c22-7273-4458-a0db-71ec3fbf3638/1055-1064

The extra query is triggered here: https://github.com/rails/rails/blob/83a6ac3fee8fd538ce7e0088913ff54f0f9bcb6f/activerecord/lib/active_record/relation/finder_methods.rb#L440

Which mean that in such case `load_async` would actually trigger a query in the main thread. 

There *might* be a way to refactor this out, but I chose to prioritize a smaller more comprehensible PR.